### PR TITLE
feat: add option to configure span name

### PR DESCRIPTION
**Pull request by fbogsany**
This adds a configuration option for MySQL instrumentation to control the span name. At present, we default to expensive client-side parsing of the expression to determine the statement type, however this is contrary to recommendations in the Semantic Conventions.

This adds 3 options, defaulting to the current behaviour: `statement_type`, `db_name` and `db_operation_and_name`. The `db_name` option extracts the database name from the `query_options`, if available. The `db_operation_and_name` option additionally looks for `db.operation` in the `OpenTelemetry::Instrumentation::Mysql2.attributes` (from the current context). This allows callers to add `db.operation` to the context, if they know it.

In all cases, if no useful span name can be determined, we default to `mysql`.

### TODO
- [x] Add tests for new config option

### CODE REVIEW

**Review by robertlaurin — APPROVED** [review]

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb
Comment (ericmustin) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb:33]: not exactly the most descriptive configuration option but i don't really have any better ideas

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb
Comment (plantfansam) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:103]: Not a new problem and hence not a blocker, but worth flagging that `database_name` could be an empty string (see code [here](https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/222/files#diff-09613a195baead8b773b7041065d5878ed88c7742a035db6e9bee436cbc05c0eR119)).

**Review by ericmustin — APPROVED** [review]
lgtm, if i recall correctly there wasn't a lot of intention when i wrote this originally, was mostly just c/p from dd-trace-rb, so makes sense to have some non performant code paths

File: instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb
Comment (plantfansam) [inline instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb:354]: Could also stub out the `database_name` method in the test, but that is less good than what's proposed in this TODO 🤷

File: instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb
Comment (plantfansam) [inline instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb:358]: Is this meant to nest under line 343? Could also just update test language to read `when span_name=db_name and db name is nil` or something.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb
Comment (plantfansam) [inline instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb:33]: Something feels weird to me about the `db_operation_and_name` option. I wonder if the option should communicate that it's reading off the `Mysql2.attributes` hash, somehow?  And writing that, maybe the option should _just_ defer to a `Mysql2.attributes` attribute and not include the `db_name`? And if users want to also inject db name, there could be another option for that? `%I[statement db_name mysql_attr enriched_mysql_attr]` or something. To be clear, I'm just musing, and don't think this is a blocker.

**Review by plantfansam — APPROVED** [review]
This seems totally reasonable to me. I have a couple thoughts inline, but nothing I feel strongly enough about to block on.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb
Comment (fbogsany) [reply instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb:33]: The argument for `"#{db_operation} #{db_name}"` is [that that's what the spec suggests](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md):
> If `db.sql.table` is not available due to its semantics, the span SHOULD be named `<db.operation> <db.name>`.

Of course, that raises the question: can we get the table somehow without parsing?
> SQL spans SHOULD be named the following way, unless the statement is known to be of low cardinality: `<db.operation> <db.name>.<db.sql.table>`, provided that `db.operation` and `db.sql.table` are available.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb
Comment (fbogsany) [reply instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:103]: Empty string or `nil`? The `nil` case is handled below with `|| 'mysql'`. I'm not sure which code you linked to.

File: instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb
Comment (fbogsany) [reply instrumentation/mysql2/test/opentelemetry/instrumentation/mysql2/instrumentation_test.rb:358]: It is meant to nest, yes - this lets me ... `let(:database) { nil }`, overriding the default in the `describe 'tracing'` block on line 36.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb
Comment (plantfansam) [reply instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/instrumentation.rb:33]: Ah, RTFS 😄 - makes sense.

File: instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb
Comment (plantfansam) [reply instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb:103]: I was reading the code wrong, but was looking at the [database_name definition](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb#L117-L120C14).  It can result in a `nil` (`(nil)&.to_s => nil`), which is handled, here.

**Pull request merged**
